src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -2082,30 +2157,77 @@ private static bool IsOffsetContained(int offset, int start, int length)
+            var sizeOfArray = _stack.Pop();
+            var eeTypeDesc = _compilation.TypeSystemContext.SystemModule.GetKnownType("System", "EETypePtr");
+            var arguments = new StackEntry[] { new LoadExpressionEntry(StackValueKind.ValueType, "eeType", GetEETypePointerForTypeDesc(arrayType, true), eeTypeDesc.MakePointerType()), sizeOfArray };
+            //TODO: call GetNewArrayHelperForType from JitHelper.cs (needs refactoring)
+            PushNonNull(CallRuntime(_compilation.TypeSystemContext, RuntimeExport, "RhNewArray", arguments, arrayType));
+        }
+
+        private LLVMValueRef ArrayBaseSize()
+        {
+            return BuildConstInt32(2 * _compilation.NodeFactory.Target.PointerSize);
         }
 
         private void ImportLoadElement(int token)
         {
+            ImportLoadElement(ResolveTypeToken(token));
         }
 
         private void ImportLoadElement(TypeDesc elementType)
         {
+            StackEntry index = _stack.Pop();
+            StackEntry arrayReference = _stack.Pop();
+            var nullSafeElementType = elementType ?? GetWellKnownType(WellKnownType.Object);
+            PushLoadExpression(GetStackValueKind(nullSafeElementType), "ldelem", GetElementAddress(index.ValueAsInt32(_builder, true), arrayReference.ValueAsType(LLVM.PointerType(LLVM.Int8Type(), 0), _builder), nullSafeElementType), nullSafeElementType);
         }
 
         private void ImportStoreElement(int token)
         {
+            ImportStoreElement(ResolveTypeToken(token));
         }
 
         private void ImportStoreElement(TypeDesc elementType)
         {
+            StackEntry value = _stack.Pop();
+            StackEntry index = _stack.Pop();
+            StackEntry arrayReference = _stack.Pop();
+            var nullSafeElementType = elementType ?? GetWellKnownType(WellKnownType.Object);
+            LLVMValueRef elementAddress = GetElementAddress(index.ValueAsInt32(_builder, true), arrayReference.ValueAsType(LLVM.PointerType(LLVM.Int8Type(), 0), _builder), nullSafeElementType);
+            CastingStore(elementAddress, value, nullSafeElementType);
         }
 
         private void ImportLoadLength()
         {
+            StackEntry arrayReference = _stack.Pop();
+            LLVMValueRef lengthPtr = LLVM.BuildGEP(_builder, arrayReference.ValueAsType(LLVM.PointerType(LLVM.Int8Type(), 0), _builder), new LLVMValueRef[] { BuildConstInt32(_compilation.NodeFactory.Target.PointerSize) }, "arrayLength");
+            LLVMValueRef castLengthPtr = LLVM.BuildPointerCast(_builder, lengthPtr, LLVM.PointerType(LLVM.Int32Type(), 0), "castArrayLength");
+            PushLoadExpression(StackValueKind.Int32, "arrayLength", castLengthPtr, GetWellKnownType(WellKnownType.Int32));
         }
 
         private void ImportAddressOfElement(int token)
         {
+            TypeDesc elementType = ResolveTypeToken(token);
+            var byRefElement = elementType.MakeByRefType();
```

Comment:
Is this a problem for pointer arrays?

Reply:
I'm not sure I understand your question. Whats wrong with making a pointer type into a reference to a pointer type? I can add a test for a pointer array if you want.

Reply:
I'd thought byrefs to pointers weren't legal, but on further testing, I was wrong. This should be fine.